### DISCUSSION TITLE
Smithy and Scarango

### DISCUSSION THREAD

**Discussion by zetashift** [disneystreaming/smithy4s#474] [Q&A] [answered]
So for a personal project I'm using ArangoDB(it's like DynamoDB but with graphs) and the scala driver https://github.com/outr/scarango

A `Document` in scarango is defined for example as follow(from the README):

```scala
// Case class to represent a person collection
case class Person(name: String, age: Int, _id: Id[Person] = Person.id()) extends Document[Person]

// We use the companion object to represent additional information about the collection
object Person extends DocumentModel[Person] {
  override implicit val rw: RW[Person] = ccRW

  val name: Field[String] = field("name")
  val age: Field[Int] = field("age")

  override def indexes: List[Index] = List(
    name.index.persistent()
  )

  override val collectionName: String = "people"
}

// We represent our entire database here referencing all collections
object Database extends Graph("example") {
  val people: DocumentCollection[Person] = vertex[Person](Person)
}
```

So I have made a `Person` API where smithy4s generates the case classes, but I'd need some way to this generated person to be a `Document[Person]` and `DocumentModel[Person]`.
What would a nice way to do this be?

I guess this is more a Scala than a smithy question, so sorry if this is the wrong place to ask! I also tried reading this part : https://blog.indoorvivants.com/2022-06-11-smithy4s-fullstack-part-2#codecs
On how to go from `model -> db codec` but as a complete noob I didn't really understand it well ><

**Comment by kubukoz** [comment, marked as answer]
I think you won't be able to make the generated class extend `Document[Person]` directly, but you might be able to make an implicit conversion for that.

For `DocumentModel[Person]`, I'd advise you to try and implement `SchemaVisitor[DocumentModel]`. Still, even if it's possible, you won't be able to call e.g. `Person.name`.

This might fall out of scope for Smithy4s usage. We don't provide this level of customization of the generated code, and as far as I know it's not likely to change.

**Reply by zetashift** [reply]
Yea I totally get understand that it falls out of scope for smithy4s! I was just wondering what a nice way to encode them would be :P
I'll look into the implicit conversion and the `SchemaVisitor` thing, thank you.

**Reply by Baccata** [reply]
@kubukoz is correct in that this falls out of the scope of Smithy4s. You'd have to create you own code-generator for this, we're not really intending on making ours customisable, in the near/medium term anyway. 

Jakub is also right in that it may be possible to get derive a `fabric.rw.RW` from smithy4s schemas, via a `smithy4s.schema.SchemaVisitor`. All data types generated are accompanied by a "schema" which allows for introspecting their structures. That's how Smithy4s decouples the generated code from specific formats and libraries. 

But besides that, it looks like scarango has a lot of `D <: Document[D]` type constraints, which implies that it can only work with datatypes that extend their `Document` types, which is not something that Smithy4s can do, I'm afraid.

**Reply by Baccata** [reply]
That being said, Smithy makes it reasonably easy to create code generators on top of it, so you could have a bespoke code generator tool that'd interop with Scarango. Obviously this is a reasonably large investment compared to what you were probably hoping to find

**Reply by zetashift** [reply]
Thanks for the thoughtful answers! The whole SchemaVisitor and a bespoke tool seems a bit too difficult for me, but since this is all for a personal project, I might try and see if I can learn a thing or two.
